x/nft-transfer/keeper/packet.go

```diff
@@ -0,0 +1,198 @@
+package keeper
+
+import (
+	"strings"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	clienttypes "github.com/cosmos/ibc-go/v5/modules/core/02-client/types"
+	channeltypes "github.com/cosmos/ibc-go/v5/modules/core/04-channel/types"
+	"github.com/crypto-org-chain/chain-main/v4/x/nft-transfer/types"
+)
+
+// refundPacketToken will unescrow and send back the tokens back to sender
+// if the sending chain was the source chain. Otherwise, the sent tokens
+// were burnt in the original send so new tokens are minted and sent to
+// the sending address.
+func (k Keeper) refundPacketToken(ctx sdk.Context, packet channeltypes.Packet, data types.NonFungibleTokenPacketData) error {
+	sender, err := sdk.AccAddressFromBech32(data.Sender)
+	if err != nil {
+		return err
+	}
+
+	classTrace := types.ParseClassTrace(data.ClassId)
+	voucherClassID := classTrace.IBCClassID()
+
+	isAwayFromOrigin := types.IsAwayFromOrigin(packet.GetSourcePort(),
+		packet.GetSourceChannel(), data.ClassId)
+
+	escrowAddress := types.GetEscrowAddress(packet.GetSourcePort(), packet.GetSourceChannel())
+
+	if isAwayFromOrigin {
```

Comment:
for this if, it may be helpful to add those comments: `// unescrow token back to sender` and `// we are sink chain, mint voucher back to sender`

Reply:
Done